lib/Notification/AppHint.php

```diff
@@ -0,0 +1,82 @@
+<?php
+/**
+ * @copyright Copyright (c) 2019 Julius Härtl <jus@bitgrid.net>
+ *
+ * @author Julius Härtl <jus@bitgrid.net>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\FirstRunWizard\Notification;
+
+use OCP\App\IAppManager;
+use OCP\IConfig;
+use OCP\IGroupManager;
+use OCP\Notification\IManager as INotificationManager;
+
+class AppHint  {
+
+	/** @var INotificationManager */
+	protected $notificationManager;
+
+	/** @var IGroupManager */
+	protected $groupManager;
+
+	/** @var IAppManager */
+	protected $appManager;
+
+	/** @var IConfig */
+	protected $config;
+
+	public function __construct(INotificationManager $notificationManager, IGroupManager $groupManager, IAppManager $appManager, IConfig $config, $userId) {
+		$this->notificationManager = $notificationManager;
+		$this->groupManager = $groupManager;
+		$this->appManager =$appManager;
+		$this->config = $config;
+		$this->userId = $userId;
+	}
+
+	public function sendAppHintNotifications(): void {
+		if ($this->userId !== null && $this->groupManager->isAdmin($this->userId)) {
+			$this->sendNotification('calendar', $this->userId);
+			$this->sendNotification('contacts', $this->userId);
+			$this->sendNotification('mail', $this->userId);
+			$this->sendNotification('spreed', $this->userId);
+			$this->config->setUserValue($this->userId, 'firstrunwizard', 'apphint', 'yes');
+		}
+	}
+
+	protected function sendNotification(string $app, string $user): void {
+		$notification = $this->notificationManager->createNotification();
+		$notification->setApp('firstrunwizard')
+			->setSubject('apphint-'. $app)
+			->setObject('app', $app)
+			->setUser($user);
+
+		if ($this->config->getUserValue($this->userId, 'firstrunwizard', 'apphint') !== 'yes' && $this->notificationManager->getCount($notification) === 0) {
+			$notification->setDateTime(new \DateTime());
+			$this->notificationManager->notify($notification);
+		}
+		if ($this->notificationManager->getCount($notification) === 0) {
+			return;
+		}
+
+		if ($this->appManager->isEnabledForUser($app)) {
+			$this->notificationManager->markProcessed($notification);
```

Comment:
Why do you notify and mark as processed if the app is enabled already? Isn't there a way to just not notify in this case?

I know that his one marks notifcations as processed when the app was enabled in the meantime, but maybe this can be made less invasive, because now this code is called on every single request that triggers the "loadAdditionalScripts"

Reply:
@MorrisJobke I've adjusted this slightly. The notification sending method will only be triggered once now and the app listens for app enable events to dismiss existing notifications in that case.